### ISSUE TITLE
Test failures on 1.6

### ISSUE THREAD

**Issue by mcabbott** [invenia/NamedDims.jl#152]
PkgEval shows a crash with "signal (4): Illegal instruction", as well as some problem with `@test_modern` I think?
https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/56ff37a_vs_ebcc603/NamedDims.1.6.0-beta1-73f862b30f.log

The crash is from this testset:
https://github.com/invenia/NamedDims.jl/blob/master/test/cat.jl#L64
and pasting it into the REPL seems to cause the same crash. 

But setting `f, d = hcat, 2` and then pasting the contents of the outermost loop, everything passes.

**Comment by oxinabox** [comment]
> crash with "signal (4): Illegal instruction",

Sounds like a julia bug.
It should be impossible to get a hard-crash unless you are doing `ccall` or `@inbounds` etc

>  some problem with `@test_modern`

`@test_modern` is is a very simple macro wrapping `@test` in a condition to make sure it only runs on fairly recent versions of julia.
Seems hard for that to break.
But not impossible if it is some weird namespacing thing i guess

**Comment by mcabbott** [comment]
The crash does not happen on 1.7.0-DEV.211 (2021-01-04), for me. 

Re the macro, I think the offending line is 
https://github.com/invenia/NamedDims.jl/blob/master/test/name_core.jl#L77
which has nested macros & dollar signs --- easy to mess up, but still weird that it's changed.

**Comment by oxinabox** [comment]
> Re the macro, I think the offending line is
https://github.com/invenia/NamedDims.jl/blob/master/test/name_core.jl#L77
which has nested macros & dollar signs --- easy to mess up, but still weird that it's changed.

Yeah, if it is changed it seems likely to be a mistake.
So it would be good to try and build a MWE and open an issue on JuliaLang/julia, so it can be fixed before 1.6 comes out.

Breakages in NamedDims.jl have often been ignored for PkgEval purposes because we test number of allocations which julia doesn't garentee (which seems legit), so if we are not careful this could be overlooked.

Are you able to try and find a MWE for that?

**Comment by mcabbott** [comment]
I got it down to this:
```
  | | |_| | | | (_| |  |  Version 1.6.0-beta1 (2021-01-08)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> using NamedDims, Test

julia> for (f, d) in zip((vcat, hcat), (1, 2))
           a = [10 20; 31 40]
           nda = NamedDimsArray(a, (:x, :y))
           @testset "$f" begin
               @test f(nda) == cat(nda; dims=d)
           end
       end
Test Summary: | Pass  Total
vcat          |    1      1

signal (4): Illegal instruction: 4
in expression starting at REPL[2]:1
_zip_iterate_all at ./iterators.jl:361 [inlined]
iterate at ./iterators.jl:351 [inlined]
top-level scope at ./REPL[2]:5
jl_toplevel_eval_flex at /Applications/Julia-1.6.app/Contents/Resources/julia/lib/julia/libjulia-int
```

**Comment by oxinabox** [comment]
Yes, that replicates for me on Mac

**Comment by mcabbott** [comment]
The crash is now fixed, by #39236.  The macro interpolation problem remains.
